src/CMakeLists.txt

```diff
@@ -360,7 +360,7 @@ set(HALIDE_H "${Halide_BINARY_DIR}/include/Halide.h")
 set(LICENSE_PATH "${Halide_SOURCE_DIR}/LICENSE.txt")
 add_custom_command(OUTPUT "${Halide_BINARY_DIR}/include/Halide.h"
                    COMMAND ${CMAKE_COMMAND} -E make_directory "$<SHELL_PATH:${Halide_BINARY_DIR}/include>"
-                   COMMAND build_halide_h "$<SHELL_PATH:${LICENSE_PATH}>" ${HEADER_FILES} > "$<SHELL_PATH:${HALIDE_H}>"
+                   COMMAND $<TARGET_FILE:build_halide_h> "$<SHELL_PATH:${LICENSE_PATH}>" ${HEADER_FILES} > "$<SHELL_PATH:${HALIDE_H}>"
```

Comment:
I don't think this is necessary with the crosscompiling emulator change above.

Reply:
Alas, maybe it shouldn't be, but it still is :-/

Reply:
Are you sure about this? I just tested it locally with removing the `$<TARGET_FILE:...>` around `binary2cpp` and it worked fine. Maybe you changed the toolchain file without first deleting the build directory?

Reply:
Here's my repro script:

```
#!/bin/bash

set -ex

rm -rf build
git diff
cmake -G Ninja -S . -B build --toolchain cmake/toolchain.linux-x64-asan.cmake -DLLVM_DIR=/usr/lib/llvm-15/lib/cmake/llvm -DHalide_SHARED_LLVM=YES -DCMAKE_BUILD_TYPE=Release
cmake --build build --target HalideIncludes --verbose
```

Here's its output, note the git diff:

```
(default) reinking@penguin:~/Halide$ ./repro.sh 
+ rm -rf build
+ git diff
diff --git a/cmake/toolchain.linux-x64-asan.cmake b/cmake/toolchain.linux-x64-asan.cmake
index 0c6c75120..c14ad3b14 100644
--- a/cmake/toolchain.linux-x64-asan.cmake
+++ b/cmake/toolchain.linux-x64-asan.cmake
@@ -34,7 +34,7 @@ set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
 set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
 
 # Set to empty string to indicate the resulting binaries can be natively executed
-set(CMAKE_CROSSCOMPILING_EMULATOR)
+set(CMAKE_CROSSCOMPILING_EMULATOR /usr/bin/env)
 
 # A few downstream build rules need to be able to sniff these:
 # - ASAN means we can't run with large stacks, so some things will overflow
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index dd86add7a..c31e37c32 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -360,7 +360,7 @@ set(HALIDE_H "${Halide_BINARY_DIR}/include/Halide.h")
 set(LICENSE_PATH "${Halide_SOURCE_DIR}/LICENSE.txt")
 add_custom_command(OUTPUT "${Halide_BINARY_DIR}/include/Halide.h"
                    COMMAND ${CMAKE_COMMAND} -E make_directory "$<SHELL_PATH:${Halide_BINARY_DIR}/include>"
-                   COMMAND $<TARGET_FILE:build_halide_h> "$<SHELL_PATH:${LICENSE_PATH}>" ${HEADER_FILES} > "$<SHELL_PATH:${HALIDE_H}>"
+                   COMMAND build_halide_h "$<SHELL_PATH:${LICENSE_PATH}>" ${HEADER_FILES} > "$<SHELL_PATH:${HALIDE_H}>"
                    DEPENDS build_halide_h "${LICENSE_PATH}" ${HEADER_FILES}
                    WORKING_DIRECTORY "${CMAKE_CURRENT_LIST_DIR}"
                    VERBATIM)
+ cmake -G Ninja -S . -B build --toolchain cmake/toolchain.linux-x64-asan.cmake -DLLVM_DIR=/usr/lib/llvm-15/lib/cmake/llvm -DHalide_SHARED_LLVM=YES -DCMAKE_BUILD_TYPE=Release
-- The C compiler identification is Clang 15.0.0
-- The CXX compiler identification is Clang 15.0.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/lib/llvm-15/lib/cmake/llvm/../../..//bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/lib/llvm-15/lib/cmake/llvm/../../..//bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Halide 'host' platform triple:   x86-64-linux
-- Halide 'cmake' platform triple:  x86-64-linux
-- Halide default AOT target:       host
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Could NOT find OpenGL (missing: OPENGL_opengl_LIBRARY OPENGL_glx_LIBRARY OPENGL_INCLUDE_DIR) 
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
-- Performing Test HAVE_FFI_CALL
-- Performing Test HAVE_FFI_CALL - Success
-- Found FFI: /usr/lib/x86_64-linux-gnu/libffi.so  
-- Performing Test Terminfo_LINKABLE
-- Performing Test Terminfo_LINKABLE - Success
-- Found Terminfo: /usr/lib/x86_64-linux-gnu/libtinfo.so  
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11") 
-- Found LibXml2: /usr/lib/x86_64-linux-gnu/libxml2.so (found version "2.9.10") 
-- Found LLVM 15.0.0
-- Using LLVMConfig.cmake in: /usr/lib/llvm-15/lib/cmake/llvm
-- Using ClangConfig.cmake in: /usr/lib/llvm-15/lib/cmake/clang
-- Enabling AArch64 backend
-- Enabling AMDGPU backend
-- Enabling ARM backend
-- Enabling Hexagon backend
-- Enabling Mips backend
-- Enabling NVPTX backend
-- Enabling PowerPC backend
-- Enabling RISCV backend
-- Enabling WebAssembly backend
-- Enabling X86 backend
-- Using LLDConfig.cmake in: /usr/lib/llvm-15/lib/cmake/lld
-- Compiling Halide WITH RTTI.
-- Compiling Halide WITH exceptions.
-- Found JPEG: /usr/lib/x86_64-linux-gnu/libjpeg.so (found version "62") 
-- Found PNG: /usr/lib/x86_64-linux-gnu/libpng.so (found version "1.6.37") 
-- Fetching WABT 1.0.29...
-- Looking for alloca.h
-- Looking for alloca.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for snprintf
-- Looking for snprintf - found
-- Looking for strcasecmp
-- Looking for strcasecmp - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of ssize_t
-- Check size of ssize_t - done
-- Check size of size_t
-- Check size of size_t - done
CMake Warning at build/_deps/wabt-src/CMakeLists.txt:214 (message):
  Error compiling dummy.c file


-- Found PythonInterp: /home/reinking/.venv/default/bin/python3 (found suitable version "3.9.2", minimum required is "3.5") 
-- Performing Test LINKER_HAS_FLAG_VERSION_SCRIPT
-- Performing Test LINKER_HAS_FLAG_VERSION_SCRIPT - Success
-- Building autoschedulers enabled
-- Building tests enabled
-- Skipping all performance testing because at least one Sanitizer is enabled.
-- Building internal runtime tests enabled
-- Building Python bindings enabled
-- Found Python3: /home/reinking/.venv/default/bin/python3.9 (found version "3.9.2") found components: Interpreter Development Development.Module Development.Embed 
-- pybind11 v2.6.2 
CMake Warning (dev) at /home/reinking/.venv/default/lib/python3.9/site-packages/cmake/data/share/cmake-3.23/Modules/CMakeDependentOption.cmake:89 (message):
  Policy CMP0127 is not set: cmake_dependent_option() supports full Condition
  Syntax.  Run "cmake --help-policy CMP0127" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.
Call Stack (most recent call first):
  build/_deps/pybind11-src/CMakeLists.txt:98 (cmake_dependent_option)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Performing Test HAS_FLTO_THIN
-- Performing Test HAS_FLTO_THIN - Success
CMake Warning at python_bindings/CMakeLists.txt:63 (message):
  Python tests/tutorials are not yet supported under ASAN.


-- Building tutorials enabled
-- Building docs disabled
-- Building utils enabled
-- Configuring done
-- Generating done
-- Build files have been written to: /home/reinking/Halide/build
+ cmake --build build --target HalideIncludes --verbose
[1/3] /usr/lib/llvm-15/lib/cmake/llvm/../../..//bin/clang++   -fsanitize=address -O3 -DNDEBUG -std=c++17 -MD -MT tools/CMakeFiles/build_halide_h.dir/build_halide_h.cpp.o -MF tools/CMakeFiles/build_halide_h.dir/build_halide_h.cpp.o.d -o tools/CMakeFiles/build_halide_h.dir/build_halide_h.cpp.o -c /home/reinking/Halide/tools/build_halide_h.cpp
[2/3] : && /usr/lib/llvm-15/lib/cmake/llvm/../../..//bin/clang++ -fsanitize=address -O3 -DNDEBUG -fsanitize=address tools/CMakeFiles/build_halide_h.dir/build_halide_h.cpp.o -o tools/build_halide_h   && :
[3/3] cd /home/reinking/Halide/src && /home/reinking/.venv/default/lib/python3.9/site-packages/cmake/data/bin/cmake -E make_directory /home/reinking/Halide/build/include && /usr/bin/env /home/reinking/Halide/build/tools/build_halide_h /home/reinking/Halide/LICENSE.txt AbstractGenerator.h AddAtomicMutex.h AddImageChecks.h AddParameterChecks.h AlignLoads.h AllocationBoundsInference.h ApplySplit.h Argument.h AssociativeOpsTable.h Associativity.h AsyncProducers.h AutoScheduleUtils.h BoundaryConditions.h Bounds.h BoundsInference.h BoundSmallAllocations.h Buffer.h Callable.h CanonicalizeGPUVars.h ClampUnsafeAccesses.h Closure.h CodeGen_C.h CodeGen_D3D12Compute_Dev.h CodeGen_GPU_Dev.h CodeGen_Internal.h CodeGen_LLVM.h CodeGen_Metal_Dev.h CodeGen_OpenCL_Dev.h CodeGen_OpenGLCompute_Dev.h CodeGen_Posix.h CodeGen_PTX_Dev.h CodeGen_PyTorch.h CodeGen_Targets.h CompilerLogger.h ConciseCasts.h CPlusPlusMangle.h CSE.h Debug.h DebugArguments.h DebugToFile.h Definition.h Deinterleave.h Derivative.h DerivativeUtils.h DeviceAPI.h DeviceArgument.h DeviceInterface.h Dimension.h EarlyFree.h Elf.h EliminateBoolVectors.h EmulateFloat16Math.h Error.h Expr.h ExprUsesVar.h Extern.h ExternFuncArgument.h ExtractTileOperations.h FastIntegerDivide.h FindCalls.h FindIntrinsics.h FlattenNestedRamps.h Float16.h Func.h Function.h FunctionPtr.h FuseGPUThreadLoops.h FuzzFloatStores.h Generator.h HexagonOffload.h HexagonOptimize.h ImageParam.h InferArguments.h InjectHostDevBufferCopies.h Inline.h InlineReductions.h IntegerDivisionTable.h Interval.h Introspection.h IntrusivePtr.h IR.h IREquality.h IRMatch.h IRMutator.h IROperator.h IRPrinter.h IRVisitor.h JITModule.h Lambda.h Lerp.h LICM.h LLVM_Output.h LLVM_Runtime_Linker.h LoopCarry.h Lower.h LowerParallelTasks.h LowerWarpShuffles.h MainPage.h Memoization.h Module.h ModulusRemainder.h Monotonic.h ObjectInstanceRegistry.h OffloadGPULoops.h OutputImageParam.h ParallelRVar.h Param.h Parameter.h ParamMap.h PartitionLoops.h Pipeline.h Prefetch.h Profiling.h PurifyIndexMath.h PythonExtensionGen.h Qualify.h Random.h RDom.h Realization.h RealizationOrder.h RebaseLoopsToZero.h Reduction.h RegionCosts.h RemoveDeadAllocations.h RemoveExternLoops.h RemoveUndef.h runtime/HalideBuffer.h runtime/HalideRuntime.h Schedule.h ScheduleFunctions.h Scope.h SelectGPUAPI.h Simplify.h SimplifyCorrelatedDifferences.h SimplifySpecializations.h SkipStages.h SlidingWindow.h Solve.h SplitTuples.h StmtToHtml.h StorageFlattening.h StorageFolding.h StrictifyFloat.h Substitute.h Target.h ThreadPool.h Tracing.h TrimNoOps.h Tuple.h Type.h UnifyDuplicateLets.h UniquifyVariableNames.h UnpackBuffers.h UnrollLoops.h UnsafePromises.h Util.h Var.h VectorizeLoops.h WasmExecutor.h WrapCalls.h > /home/reinking/Halide/build/include/Halide.h
```